scene/gui/text_edit.cpp

```diff
@@ -1019,17 +1019,43 @@ void TextEdit::_notification(int p_what) {
 					const String &str = wrap_rows[line_wrap_index];
 					int char_margin = xmargin_beg - cursor.x_ofs;
 
-					int ofs_readonly = 0;
 					int ofs_x = 0;
+					int ofs_y = 0;
 					if (readonly) {
-						ofs_readonly = cache.style_readonly->get_offset().y / 2;
 						ofs_x = cache.style_readonly->get_offset().x / 2;
+						ofs_x -= cache.style_normal->get_offset().x / 2;
+						ofs_y = cache.style_readonly->get_offset().y / 2;
+					} else {
+						ofs_y = cache.style_normal->get_offset().y / 2;
 					}
 
-					int ofs_y = (i * row_height + cache.line_spacing / 2) + ofs_readonly;
+					ofs_y += i * row_height + cache.line_spacing / 2;
 					ofs_y -= cursor.wrap_ofs * row_height;
 					ofs_y -= get_v_scroll_offset() * row_height;
 
+					int top_limit_y = 0;
+					int bottom_limit_y = get_size().height;
+					if (readonly) {
+						top_limit_y += cache.style_readonly->get_margin(SIDE_TOP);
+						bottom_limit_y -= cache.style_readonly->get_margin(SIDE_BOTTOM);
+					} else {
+						top_limit_y += cache.style_normal->get_margin(SIDE_TOP);
+						bottom_limit_y -= cache.style_normal->get_margin(SIDE_BOTTOM);
+					}
```

Comment:
Might be worth pre-calculating this outside of the loops.

Reply:
Done and pushed.